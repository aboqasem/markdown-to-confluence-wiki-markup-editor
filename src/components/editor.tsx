import { closeBrackets, closeBracketsKeymap } from "@codemirror/autocomplete";
import { defaultKeymap, history, historyKeymap } from "@codemirror/commands";
import {
  bracketMatching,
  defaultHighlightStyle,
  indentOnInput,
  syntaxHighlighting,
} from "@codemirror/language";
import { highlightSelectionMatches, searchKeymap } from "@codemirror/search";
import { EditorState, Extension } from "@codemirror/state";
import {
  EditorView,
  drawSelection,
  highlightActiveLine,
  highlightActiveLineGutter,
  highlightSpecialChars,
  keymap,
  lineNumbers,
} from "@codemirror/view";
import {
  CreateCodeMirrorProps,
  createCodeMirror,
  createEditorControlledValue,
} from "solid-codemirror";
import { Accessor, ComponentProps, splitProps } from "solid-js";

export const basicSetup = [
  lineNumbers(),
  highlightActiveLineGutter(),
  highlightSpecialChars(),
  history(),
  drawSelection(),
  EditorState.allowMultipleSelections.of(true),
  indentOnInput(),
  syntaxHighlighting(defaultHighlightStyle, { fallback: true }),
  bracketMatching(),
  closeBrackets(),
  highlightActiveLine(),
  highlightSelectionMatches(),
  keymap.of([...closeBracketsKeymap, ...defaultKeymap, ...searchKeymap, ...historyKeymap]),
];

export const baseTheme = EditorView.baseTheme({
  "&": {
    height: "100%",
    width: "100%",
  },
});

export function Editor(
  props: ComponentProps<"div"> &
    Pick<CreateCodeMirrorProps, "onValueChange"> & {
      value?: string | Accessor<string>;
      extensions?: (Extension | Accessor<Extension | undefined>)[];
    },
) {
  const [componentProps, divProps] = splitProps(props, ["value", "onValueChange", "extensions"]);

  const {
    editorView,
    createExtension,
    ref: editorRef,
  } = createCodeMirror({
    value:
      typeof componentProps.value === "function" ? componentProps.value() : componentProps.value,
    onValueChange: componentProps.onValueChange,
  });

  if (typeof componentProps.value === "function") {
    createEditorControlledValue(editorView, componentProps.value);
  }

  if (componentProps.extensions) {
    for (const extension of componentProps.extensions) {
      createExtension(extension);
    }
  }

  return <div {...divProps} ref={editorRef} />;
}
