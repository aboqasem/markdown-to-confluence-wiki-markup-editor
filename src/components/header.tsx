export function Header() {
  return (
    <header class="md:flex md:items-center md:justify-between bg-zinc-900 p-4">
      <div class="min-w-0 flex-1">
        <h2 class="text-2xl font-medium leading-7 text-white sm:truncate sm:text-3xl sm:tracking-tight">
          Markdown to Confluence Wiki Markup Editor
        </h2>
      </div>
    </header>
  );
}
