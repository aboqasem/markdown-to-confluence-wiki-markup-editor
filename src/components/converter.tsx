import { type ComponentProps, createSignal } from "solid-js";
import { Editor } from "~/components/editor";
import { baseTheme } from "~/extensions/base-theme";
import { lazyBasicSetup } from "~/extensions/basic-setup.lazy";
import { lazyMarkdown } from "~/extensions/lang-markdown.lazy";
import { md2Cwm } from "~/renderers/md2cwm";

export type ConverterProps = ComponentProps<"div">;

export function Converter(props: ConverterProps) {
  const inputEditorId = `${props.id}_input-editor`;
  const outputEditorId = `${props.id}_output-editor`;

  const [input, setInput] = createSignal("");

  const output = () => md2Cwm(input());

  return (
    <div
      {...props}
      data-testid={props.id}
      class="flex grow overflow-auto p-4 justify-between gap-4"
    >
      <Editor
        id={inputEditorId}
        data-testid={inputEditorId}
        class="size-full overflow-scroll"
        value={input}
        onValueChange={setInput}
        extensions={[baseTheme]}
        lazyExtensions={[lazyBasicSetup, lazyMarkdown]}
      />

      <hr class="border h-full border-gray-200" />

      <Editor
        id={outputEditorId}
        data-testid={outputEditorId}
        class="size-full overflow-scroll"
        value={output}
        extensions={[baseTheme]}
        lazyExtensions={[lazyBasicSetup]}
      />
    </div>
  );
}
