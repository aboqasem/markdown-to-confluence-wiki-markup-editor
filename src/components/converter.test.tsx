import { render } from "@solidjs/testing-library";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import { Converter } from "~/components/converter";

function assertInstanceOf<T>(value: unknown, type: new () => T): T {
  chai.assert.instanceOf(value, type);

  return value as T;
}

describe("Converter", () => {
  const converterId = "converter";
  const converterInputId = `${converterId}_input-editor`;
  const converterOutputId = `${converterId}_output-editor`;
  let rendered: ReturnType<typeof render>;
  let inputEditor: HTMLDivElement;
  let outputEditor: HTMLDivElement;

  beforeEach(() => {
    rendered = render(() => <Converter id={converterId} />);
    inputEditor = assertInstanceOf(
      rendered.getByTestId(converterInputId).querySelector(".cm-content"),
      HTMLDivElement,
    );
    outputEditor = assertInstanceOf(
      rendered.getByTestId(converterOutputId).querySelector(".cm-content"),
      HTMLDivElement,
    );
  });

  it("converts markdown to confluence wiki markup", async () => {
    const input = `
# Hello World

- This is a list item
- This is another list item

> This is a blockquote

\`\`\`js
// This is a code block
console.log("Hello World");
\`\`\`
`;

    await userEvent.type(inputEditor, input, { delay: null });

    expect(outputEditor).matchSnapshot();
  });
});
