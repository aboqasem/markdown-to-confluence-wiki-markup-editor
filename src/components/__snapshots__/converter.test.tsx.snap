// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Converter > converts markdown to confluence wiki markup 1`] = `
<div
  aria-multiline="true"
  autocapitalize="off"
  autocorrect="off"
  class="cm-content"
  contenteditable="true"
  role="textbox"
  spellcheck="false"
  style="tab-size: 4;"
  translate="no"
>
  <div
    class="cm-line"
  >
    h1. Hello World
  </div>
  <div
    class="cm-line"
  >
    * This is a list item
  </div>
  <div
    class="cm-line"
  >
    * This is another list item
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    bq. This is a blockquote
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
  <div
    class="cm-line"
  >
    {code:language=js|linenumbers=true|collapse=false}
  </div>
  <div
    class="cm-line"
  >
    // This is a code block
  </div>
  <div
    class="cm-line"
  >
    console.log("Hello World");
  </div>
  <div
    class="cm-line"
  >
    {code}
  </div>
  <div
    class="cm-line"
  >
    <br />
  </div>
</div>
`;
